### ISSUE TITLE
Code to visualize predicted 3DIS masks individually

### ISSUE THREAD

**Issue by Vinit2244** [yunshin/SphericalMask#19]
## Steps
> Note: This visualisation follows from the setup done in #15.
> This code is only for visualising either scannetv2 3DIS or custom scene inference based on scannev2 classes as described in #17.

1. Create a new file named **`vis_scannetv2_ind.py`** inside the **`SphericalMask/visualization/`** directory.
2. Copy the code provided below into this file. 
3. Open a terminal, navigate to the **`SphericalMask/`** directory (make sure your environment is active), and execute the following command to run the script.

```bash
# This is a sample command you can update the arguments being passed according to your scene and utility
python3 visualization/vis_scannetv2_ind.py \
    --data_root dataset/scannetv2 \
    --scene_name scene0000_00 \
    --prediction_path ./results \
    --split test \
    --filter_min_score \
    --min_score 0.5 \
    --filter_min_points \
    --min_points 100 \
    --apply_wta
# FLAGS Explained:
# --filter_min_score: Enable score filtering (removes low confidence objects)
# --min_score: Set the threshold for the above flag
# --filter_min_points: Enable point count filtering (removes tiny noise fragments)
# --min_points: Set the threshold for minimum points
# --apply_wta: Enable "Winner-Takes-All" (per-vertex) overlap resolution (other option available is --apply_nms - non-maximum suppression)

# Visualise
cd visualization/pyviz3d

# Start local server
python3 -m http.server 6008

# Open in browser
http://0.0.0.0:6008
```

---

## `vis_scannetv2_ind.py` file content:
```python
import os
import numpy as np
import pyviz3d.visualizer as viz
import torch
import os.path as osp
import argparse

# Detectron2 Color Palette
COLOR_DETECTRON2 = (
    np.array(
        [
            0.000, 0.447, 0.741, 0.850, 0.325, 0.098, 0.929, 0.694, 0.125, 0.494, 0.184, 0.556, 0.466, 0.674, 0.188, 0.301,
            0.745, 0.933, 0.635, 0.078, 0.184, 0.600, 0.600, 0.600, 1.000, 0.000, 0.000, 1.000, 0.500, 0.000, 0.749, 0.749,
            0.000, 0.000, 1.000, 0.000, 0.000, 0.000, 1.000, 0.667, 0.000, 1.000, 0.333, 0.333, 0.000, 0.333, 0.667, 0.000,
            0.333, 1.000, 0.000, 0.667, 0.333, 0.000, 0.667, 0.667, 0.000, 0.667, 1.000, 0.000, 1.000, 0.333, 0.000, 1.000,
            0.667, 0.000, 1.000, 1.000, 0.000, 0.000, 0.333, 0.500, 0.000, 0.667, 0.500, 0.000, 1.000, 0.500, 0.333, 0.000,
            0.500, 0.333, 0.333, 0.500, 0.333, 0.667, 0.500, 0.333, 1.000, 0.500, 0.667, 0.000, 0.500, 0.667, 0.333, 0.500,
            0.667, 0.667, 0.500, 0.667, 1.000, 0.500, 1.000, 0.000, 0.500, 1.000, 0.333, 0.500, 1.000, 0.667, 0.500, 1.000,
            1.000, 0.500, 0.000, 0.333, 1.000, 0.000, 0.667, 1.000, 0.000, 1.000, 1.000, 0.333, 0.000, 1.000, 0.333, 0.333,
            1.000, 0.333, 0.667, 1.000, 0.333, 1.000, 1.000, 0.667, 0.000, 1.000, 0.667, 0.333, 1.000, 0.667, 0.667, 1.000,
            0.667, 1.000, 1.000, 1.000, 0.000, 1.000, 1.000, 0.333, 1.000, 1.000, 0.667, 1.000, 0.500, 0.000, 0.000, 0.667,
            0.000, 0.000, 0.833, 0.000, 0.000, 1.000, 0.000, 0.000, 0.000, 0.167, 0.000, 0.000, 0.500, 0.000, 0.000, 0.667,
            0.000, 0.000, 0.833, 0.000, 0.000, 1.000, 0.000, 0.000, 0.000, 0.167, 0.000, 0.000, 0.500, 0.000, 0.000, 0.667,
            0.000, 0.000, 0.833, 0.000, 0.000, 1.000, 0.143, 0.143, 0.143, 0.857, 0.857, 0.857
        ]
    )
    .astype(np.float32)
    .reshape(-1, 3)
    * 255
)

SEMANTIC_IDX2NAME = {
    1: "wall", 2: "floor", 3: "cabinet", 4: "bed", 5: "chair", 6: "sofa", 7: "table",
    8: "door", 9: "window", 10: "bookshelf", 11: "picture", 12: "counter", 14: "desk",
    16: "curtain", 24: "refridgerator", 28: "shower curtain", 33: "toilet", 34: "sink",
    36: "bathtub", 39: "otherfurniture",
}

class PredictedObject:
    def __init__(self, label_id, indices, score, file_name):
        self.label_id = label_id
        self.indices = set(indices) # Set for fast intersection (NMS)
        self.score = score
        self.file_name = file_name
        self.array_indices = indices # Array for fast indexing (WTA/Viz)


def resolve_overlaps_wta(candidates, num_vertices):
    # Track who owns each vertex
    vertex_owner_idx = np.full(num_vertices, -1, dtype=int)
    vertex_max_scores = np.full(num_vertices, -1.0, dtype=float)

    for i, cand in enumerate(candidates):
        indices = cand.array_indices
        score = cand.score
        
        # Compare new score with existing scores at these vertices
        current_scores = vertex_max_scores[indices]
        better_mask = score > current_scores
        indices_to_update = indices[better_mask]
        
        if len(indices_to_update) > 0:
            vertex_owner_idx[indices_to_update] = i
            vertex_max_scores[indices_to_update] = score

    # Group back by owner
    unique_owners = np.unique(vertex_owner_idx)
    unique_owners = unique_owners[unique_owners != -1]
    
    final_results = []
    for owner_idx in unique_owners:
        cand = candidates[owner_idx]
        mask_owned = np.where(vertex_owner_idx == owner_idx)[0]
        final_results.append((cand, mask_owned))
        
    return final_results

def resolve_overlaps_nms(candidates, iou_thresh):
    # Sort by score descending
    candidates.sort(key=lambda x: x.score, reverse=True)
    kept_objects = []

    for cand in candidates:
        is_duplicate = False
        for kept in kept_objects:
            if cand.label_id == kept.label_id:
                intersection = len(cand.indices.intersection(kept.indices))
                union = len(cand.indices.union(kept.indices))
                iou = intersection / union if union > 0 else 0.0
                
                if iou > iou_thresh:
                    is_duplicate = True
                    break
        
        if not is_duplicate:
            kept_objects.append(cand)

    # For NMS, we keep the original full masks
    return [(obj, obj.array_indices) for obj in kept_objects]

def add_pred_instances_to_viz(visualizer, xyz_points, scene_name, dir, args):
    instance_file = osp.join(dir, "pred_instance", scene_name + ".txt")
    if not osp.exists(instance_file):
        print(f"Prediction file not found: {instance_file}")
        return

    # Load raw data
    with open(instance_file, "r") as f:
        parsed_lines = [line.strip().split() for line in f.readlines()]
    
    candidates = []
    num_vertices = xyz_points.shape[0]

    for parts in parsed_lines:
        if len(parts) < 3: continue
        mask_filename, label_id, score = parts[0], int(parts[1]), float(parts[2])

        if label_id not in SEMANTIC_IDX2NAME: continue

        if args.filter_min_score and score < args.min_score:
            continue

        mask_path = osp.join(dir, "pred_instance", mask_filename)
        if not osp.isfile(mask_path): continue
        
        try:
            full_mask = np.loadtxt(mask_path).astype(np.int32)
        except Exception: continue

        if full_mask.shape[0] != num_vertices: continue
        
        indices = np.where(full_mask == 1)[0]

        if args.filter_min_points and len(indices) < args.min_points:
            continue

        candidates.append(PredictedObject(label_id, indices, score, mask_filename))

    print(f"Candidates after loading and basic filtering: {len(candidates)}")

    final_items = []

    if args.apply_wta:
        # Winner Takes All (Per Vertex)
        final_items = resolve_overlaps_wta(candidates, num_vertices)
    elif args.apply_nms:
        # NMS (Intersection over Union)
        final_items = resolve_overlaps_nms(candidates, args.iou_threshold)
    else:
        # Raw (Overlapping)
        print("No overlap resolution selected. Displaying all valid candidates.")
        final_items = [(c, c.array_indices) for c in candidates]

    print(f"Final objects to visualize: {len(final_items)}")

    # Visualize
    final_items.sort(key=lambda x: x[0].score, reverse=True)

    for i, (obj, visible_indices) in enumerate(final_items):
        if len(visible_indices) == 0: continue

        cls_name = SEMANTIC_IDX2NAME.get(obj.label_id, f"c{obj.label_id}")
        short_cls = cls_name[:5]
        layer_name = f"P{i}:{short_cls}:{obj.score:.2f}"
        
        obj_xyz = xyz_points[visible_indices]
        color_val = COLOR_DETECTRON2[i % len(COLOR_DETECTRON2)]
        obj_rgb = np.tile(color_val, (obj_xyz.shape[0], 1))

        visualizer.add_points(layer_name, obj_xyz, obj_rgb, point_size=args.point_size)

def main():
    parser = argparse.ArgumentParser("ScanNetV2-Vis-Granular-Filtering")
    
    parser.add_argument("--data_root", type=str, default='dataset/scannetv2')
    parser.add_argument("--scene_name", type=str, default='scene0000_00')
    parser.add_argument("--split", type=str, default='test')
    parser.add_argument("--prediction_path", default="./results")
    parser.add_argument("--point_size", type=float, default=15.0)
    
    parser.add_argument("--filter_min_score", action='store_true', help="Enable minimum confidence score filtering")
    parser.add_argument("--filter_min_points", action='store_true', help="Enable minimum point count filtering")
    parser.add_argument("--apply_nms", action='store_true', help="Enable NMS (IoU-based) overlap resolution")
    parser.add_argument("--apply_wta", action='store_true', help="Enable Winner-Takes-All (Per-Vertex) overlap resolution")

    parser.add_argument("--min_score", type=float, default=0.5, help="Threshold for min_score filter")
    parser.add_argument("--min_points", type=int, default=100, help="Threshold for min_points filter")
    parser.add_argument("--iou_threshold", type=float, default=0.25, help="Threshold for NMS IoU")

    args = parser.parse_args()

    # Conflict check
    if args.apply_nms and args.apply_wta:
        print("WARNING: Both NMS and WTA selected. Defaulting to WTA (Winner-Takes-All).")

    v = viz.Visualizer()

    # Load Data
    file_path = f'{args.data_root}/{args.split}/{args.scene_name}_inst_nostuff.pth'
    if not os.path.exists(file_path):
        print(f"Error: File not found at {file_path}")
        return

    data = torch.load(file_path)
    if len(data) == 4:
        xyz, rgb, semantic_label, instance_label = data
    else:
        xyz, rgb = data
        semantic_label, instance_label = None, None

    xyz = xyz.astype(np.float32)
    rgb = rgb.astype(np.float32)
    rgb = (rgb + 1.0) * 127.5

    # Apply Mask Valid (GT side)
    if semantic_label is not None:
        mask_valid = (semantic_label != -100)
        xyz = xyz[mask_valid]
        rgb = rgb[mask_valid]
        semantic_label = semantic_label[mask_valid]
        instance_label = instance_label[mask_valid]

    # Visualize Predictions
    print("Processing Predicted Instances...")
    add_pred_instances_to_viz(v, xyz, args.scene_name, args.prediction_path, args)

    # Save
    output_dir = 'visualization/pyviz3d'
    os.makedirs(output_dir, exist_ok=True)
    v.save(output_dir)
    print(f"Visualization saved to {output_dir}")

if __name__ == '__main__':
    main()
```
